### PR TITLE
Turn on more CWL tests

### LINKED ISSUE #3538: Fix CWL 1.2 conformance test 35 to reliably succeed on Kubernetes
Sometimes this test fails when run on Kubernetes. It should always pass, and it should be made a non-optional part of CI.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-851)
┆Epic: CWL v1.2 support
┆Issue Number: TOIL-851

### LINKED ISSUE #3539: Fix CWL 1.2 conformance test 39 to reliably succeed on Kubernetes
Sometimes this test fails when run on Kubernetes. It should always pass, and it should be made a non-optional part of CI.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-852)
┆Epic: CWL v1.2 support
┆Issue Number: TOIL-852

### LINKED ISSUE #3540: Fix CWL 1.2 conformance test 42 to reliably succeed on Kubernetes
Sometimes this test fails when run on Kubernetes. It should always pass, and it should be made a non-optional part of CI.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-853)
┆Epic: CWL v1.2 support
┆Issue Number: TOIL-853

### LINKED ISSUE #3541: Fix CWL 1.2 conformance test 56 to reliably succeed on Kubernetes
Sometimes this test fails when run on Kubernetes. It should always pass, and it should be made a non-optional part of CI.

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-854)
┆Epic: CWL v1.2 support
┆Issue Number: TOIL-854

### PR DESCRIPTION
**Pull request by adamnovak**
## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * CWL 1.2 tests 35, 39, 42, and 56 now run (#3538, #3539, #3540, #3541)
 * Kubernetes batch system no longer runs the root job an extra time on the cluster if it is a CWL local job
 * Job descriptions track the version number of the description, for debugging job description update logic
 * Debug logging for promise tracking has been expanded
 * CWL tools which use input files where the paths are specified in the tool definition will now have their input files imported properly

Closes #3538
Closes #3539
Closes #3540
Closes #3541

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/cliOptions.rst`
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
This is a lie; I've reproduced some of these failing with complaints about the promise-fiddling job at the end of the workflow running too soon.

**Comment by adamnovak** [comment]
This can reproduce the issue with test 39:

```
FAILED=0
for TRY_NUM in {1..5} ; do
    for ITER in {1..32} ; do
        OMP_NUM_THREADS=1 TOIL_AWS_ZONE=us-west-2a TOIL_AWS_SECRET_NAME=shared-s3-credentials TOIL_APPLIANCE_SELF=quay.io/ucsc_cgl/toil:5.4.0a1-1a69dc2607c73f3b1eb707a1bab22693071c07ad-py3.6 'cwltest' '--verbose' '--tool=toil-cwl-runner' '--test=/public/home/anovak/build/toil/src/toil/test/cwl/spec_v12/conformance_tests.yaml' '--timeout=2400' '--basedir=/public/home/anovak/build/toil/src/toil/test/cwl/spec_v12' '-n=39' '-j8' '--' '--disableCaching=False' '--clean=always' '--logDebug' --setEnv OMP_NUM_THREADS '--batchSystem=kubernetes' >iter${ITER}.log 2>&1 &
    done
    
    for ITER in {1..32} ; do
        wait
        echo "Status: $?"
    done
    
    for ITER in {1..32} ; do
        grep "All tests passed" "iter${ITER}.log"
        if [[ "${?}" == "1" ]] ; then
            echo "Failure observed in iter${ITER}.log!"
            FAILED=1
            break
        fi
    done
    
    if [[ "${FAILED}" == "1" ]] ; then
        break
    fi
    echo "No failures observed in try ${TRY_NUM}"
done
```

You may want to hack the `--quiet` option in cwltoil so it actually logs at debug level.

**Comment by adamnovak** [comment]
The failures are happening because the root CWLWorkflow job is managing to get its body run twice, which messes up all the cross-references to what promises ought to be filled in when, I think.

**Comment by adamnovak** [comment]
I think the problem is that we have the same job (`CWLWorkflow`) running at the same time on Kubernetes and locally, and they are racing and generally causing trouble.

**Comment by adamnovak** [comment]
OK, the real problem was that the Kubernetes batch system was mistaking local job 0 for a non-local job and also running it on Kubernetes, leading to races and maybe the whole-workflow-runs-twice issue?

**Comment by adamnovak** [comment]
I've added a fix for test 56 to this; we weren't quite recursing right when exploring the CWL workflow looking for tool definitions and `File` objects, so we were forgetting to import files that came along as part of tool definitions instead of workflow inputs.

**Comment by adamnovak** [comment]
OK, I noticed that I had broken a test on single machine, and digging into that I noticed that the way we had been shipping Directory CWL objects around and dealing with them could never work between machines.

I didn't quite understand it to start with, but I've now changed it so we have two kinds of location paths:
* `toilfs:` for actual files, which have a FileID encoded in there
* `_:` for directories, which have a URL-safe base64-encoded JSON description of the directory structure, with files represented by their `toilfs:` paths.

When a job finishes or when we upload the initial workflow/order, we go through, get full CWL `listing` values with all files recursively uploaded, then pack those for the top-level Directories into the weird encoded `location` format. Then `ToilFsAccess` and `toil_get_file` have enough information to actually find a file, given its path relative to a directory, or to e.g. glob in a directory.

Right now `ToilFsAccess` just downloads the whole directory into a local cache as soon as someone wants to touch it, but it seems like its whole point is to present a virtual filesystem, so we could make that more efficient.

If we don't like encoding whole directory structures, we could probably write the directory structure to a file in the filestore and point to that instead. What we can't do is what we've been doing where we just use the local path corresponding to the CWL Directory's original path, and see if there's any stuff there or not.

**Comment by adamnovak** [comment]
CWL specifies that tasks be presented with a configurable-depth `listing` for each directory, and it insists on computing this when it goes to run the task. So instead of precomputing listings when we put the job in, we make sure the `listing` field is empty and only top-level Directories exist in the CWL objects, and the tasks compute the levels of listings they need via the `ToilFsAccess`.

**Comment by adamnovak** [comment]
This might not quite be ready to merge yet; I have some extraneous debugging hooks I want to rip out if it somehow passes.

File: src/toil/cwl/cwltoil.py
Comment (mr-c) [inline src/toil/cwl/cwltoil.py:2417]: What happened? `toil-cwl-runner` used to support this!

Comment (mr-c) [reply src/toil/cwl/cwltoil.py:2417]: This definitely worked for single node execution and via batch schedulers when using a shared filesystem. In other scenarios (non shared filesystems for the job store) I agree that supporting `InplaceUpdateRequirement` doesn't make sense.

File: src/toil/worker.py
Comment (mr-c) [inline src/toil/worker.py:621]: > UnboundLocalError: local variable 'fail_exit_code' referenced before assignment

File: src/toil/worker.py
Comment (mr-c) [inline src/toil/worker.py:508]:
```suggestion
            failure_exit_code = CWL_UNSUPPORTED_REQUIREMENT_EXIT_CODE
```

File: src/toil/worker.py
Comment (mr-c) [reply src/toil/worker.py:621]:
```suggestion
        return failure_exit_code
```

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:2417]: I wanted to turn this off because I'm reworking Toil to use logic that works for multiple nodes, which ends up actually keeping Directories' contents in Toil instead of on the filesystem. Right now in the implementation I have, the Directory contents are basically packed into strings that Toil passes around, meaning they can't be updated from elsewhere.

I could optimize it to check if the `single_machine` batch system is being used, or any batch system with a file job store, and then find a place for CWL to keep all its stuff in a local temp directory and support InplaceUpdateRequirement, but that seems awkward. I don't *think* the FileJobStore technically presents an interface to get a shared temp directory right now; I'm not really sure I understand how we were ensuring that files or Directories identified by paths weren't deleted when jobs ended before.

What I am hoping we might be able to do eventually is keep Directory contents in mutable files, which could be replaced from other jobs, allowing for side-effect updates from any machine to be visible on any other, with all the race conditions that you would expect in that scenario being exposed to the workflow. But I'm not sure that that semantically matches what `InplaceUpdateRequirement` is supposed to mean; Lon said it was added because people had big files that were too huge to copy, and allowing the side effects without actually being able to prevent the copy is no help to anyone.

**Comment by adamnovak** [comment]
OK, I fixed the last remaining bug where we need to skip missing files to make optional `secondaryFiles` work, and this is ready for one last review of those changes and then to be merged.

**Review by mr-c — DISMISSED** [review]
Epic PR!

I would really appreciate it if the existing support for `InplaceUpdateRequirement` on single 
machines, or with batch system backends was not removed.

File: src/toil/cwl/cwltoil.py
Comment (mr-c) [inline src/toil/cwl/cwltoil.py:151]: Check out https://github.com/common-workflow-language/cwltool/blob/dc5381e04fa2a9d18de5c78e0f8c2907919983ee/cwltool/utils.py#L85 and friends

Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:151]: I guess `CWLObjectType` could be used in place of `MutableMapping`, but this comment really refers to how the `Any` returned from `op_down` needs to be the same `Any` accepted by `op_up`, for a given call to this function. I don't know if Python's type system supports that kind of constraint.

Comment (mr-c) [reply src/toil/cwl/cwltoil.py:151]: Try `OpType= TypeVar('OpType')` and use that

**Comment by adamnovak** [comment]
Reimplementing in place update when the job store is file-backed is going to require a bunch of code changes at places that might not currently know what job store type is in use (to bypass the Directory storage system), and possibly some changes to the AbstractJobStore or FileJobStore API to give them a concept of a shared directory. I personally would want to develop it as its own ticket, so the changes can go through their own review, but I can try and pile them on top of this if we'd rather review them that way.

I think we'll still need a new ticket for me to say I'm working on though, which would then block CWL Kubernetes support until it's done.

Do we have consensus on the relative priority of CWL Kubernetes support and reimplementation of InplaceUpdateRequirement @mr-c and @DailyDreaming?

**Comment by mr-c** [comment]
As long as the in place update is restored before the next Toil release, then I'm okay.

**Comment by DailyDreaming** [comment]
@adamnovak I'm going to do a release today/tomorrow, before these are merged.  We can then work on `InplaceUpdateRequirement` prior to the release that follows.

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1249]: Suggestion:

```
    try:
        tool_id = cwl_object['id']
    except:
        tool_id = str(cwl_object)
```

->

```
tool_id = cwl_object.get('id', str(cwl_object))
```

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1258]: This sounds like it should be made into a cwltool issue and that issue linked in the comment so that future coder-lings will know if that state has changed.  I'm not sure where in the spec this is referencing and if the spec changes to address this in the future, I really won't know.

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1548]: and... ?  XD

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1545]: Is this not done with the next `pm = ToilPathMapper`call?  Or old listings are not up-to-date?

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1585]: I think if the type is a variant of `File` but the id is `toildir:` then it should error here.  That _should_ never happen.

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:1853]: This `original`/`wrapper` incantation should probably have a comment explaining itself.

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:2437]: `reqirement`

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:3089]: `couyldn't`

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:3087]: Hmmm... why was this changed from `job_name, job_params` to something more ambiguous?  Are the `initialized_job_order.items()` not all jobs?

File: src/toil/job.py
Comment (DailyDreaming) [inline src/toil/job.py:476]: Is this only for debugging?  The purpose of this isn't clear to me.

File: src/toil/cwl/cwltoil.py
Comment (DailyDreaming) [inline src/toil/cwl/cwltoil.py:3047]: When the builder operates on this in the `job.run()` step if I remember correctly.  Though I may be misremembering.

**Review by DailyDreaming — COMMENTED** [review]
The directory implementation makes sense to me and I think brings us more closely into sync with cwltool, so I'm pretty excited to have this in.

Some of the functions have some overlapping functionality that I feel could be reduced (like the traversing up and down and both and normal visit() functions), but that's probably more time than it's worth trying to reduce and more of a nit.

A few suggestions:

- Could `toilfs:` be `toilfile:` now that we have `toildir:`?
- Could `cwltoil.py` be broken into two or more files now that we're exceeding 3000 lines of code?
- I believe I commented on the relevant functions, but some of the utility functions might benefit from adding unit tests.

Note to future code archaelogists:  This PR drops support for `InplaceUpdateRequirement`, but will be added back in https://github.com/DataBiosphere/toil/pull/3652 so Toil will avoid doing a release without this support.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:151]: OK, I'll define a TypeVar for the down operation return type.

File: src/toil/job.py
Comment (adamnovak) [reply src/toil/job.py:477]: I'll rename this.

File: src/toil/job.py
Comment (adamnovak) [reply src/toil/job.py:841]: Let's go with "New job version".

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1249]: Yeah, this will work as long as the type constraints are followed. I think I meant to also maybe support lists?

Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1249]: Although it will always evaluate `str()`

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1545]: I think we don't actually need to do it after all. The path mapper and whatever stages using it is clever enough to work even without content listings.

Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1545]: I think we're using our own custom code that just makes a directory to "stage" the directory, instead of using e.g. ToilFsAccess to list its files and copy them. Seems like it might be wrong, but also there's no conformance test here that we're failing.

Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1545]: It's not that listings won't be up to date so much as it's that they might not exist or be recursive; the child Directory and File objects are only sometimes created at the CWL level, but all the contents are supposed to come along all the time.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1585]: I think it can happen when you take a `toildir:` directory and compose a path to a file within it by joining subdirectories/filenames onto its URL with `/`. That happens in some places, and I'm not sure if it can happen here or not. It might be safest to actually implement `toildir:`.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1853]: I've dropped this; I just had it so I could cram debugging in there.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:3087]: Nope! This isn't an "order" as in a sequence of jobs; it's an "order" as in instructions to run the jobs with certain arguments. I changed it to stop being misleading.

File: src/toil/job.py
Comment (adamnovak) [reply src/toil/job.py:476]: I added this for debugging because I thought we might have been forgetting to save an update or running the wrong version of a job somewhere; it turned out that wasn't the case and we were giving the same version of the job to two batch systems, but I left this in because I want to use it later when poking around the leader looking for race conditions with its weird job update system.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:128]: I'll write one.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:152]: I'm writing one for this too.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1196]: I tried calling `path_to_loc` on the already imported `cwltool.main` and it didn't exist; maybe it's not/hasn't been/isn't staying part of the API?

Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1196]: It looks like it's an internal function of `init_job_order` so we can't get at it.

File: src/toil/cwl/cwltoil.py
Comment (adamnovak) [reply src/toil/cwl/cwltoil.py:1113]: It could, but I don't think I'll have time to write it before I leave. You might have to do it, or I can do it when I get back. This one won't be as simple because we have to make up a file store and find it a place to download to. We might actually want to redesign this a bit if we really intend to test it.

**Comment by adamnovak** [comment]
OK @DailyDreaming, I've addressed most of your comments and synced this up so it's the same commit as #3652.

I split out a couple of the utility functions, but I don't think it made much of a dent in the bigness of the file. We might want to pull out ToilPathMapper, ToilFsAccess, etc. into their own files.

I also still don't have unit tests for `download_structure`, and probably won't for the next 2 weeks. If we want to ship this sooner and we need them, someone else might need to write them. Otherwise I can write them when I get back.

**Comment by adamnovak** [comment]
OK, I've fixed the problem that failed the tests on the last merge with the main development line, and added a test for `download_structure`. This *should* pass and is ready for re-review.

**Comment by adamnovak** [comment]
Actually since Lon is out, I think I might just merge this when it passes, since I have addressed the comments.

**Comment by mr-c** [comment]
Can this be closed now that #3652 is merged?

**Pull request closed without merging**
